File: backend/app/api/routes/work.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.app.api.deps import require_membership_dep
from backend.app.db import get_db
from backend.app.models import WorkItem
from backend.app.services import work_engine_service

router = APIRouter(prefix="/api/work", tags=["work"])


class SnoozeWorkIn(BaseModel):
    snoozed_until: datetime


@router.get("", dependencies=[Depends(require_membership_dep())])
def get_work_items(
    business_id: str = Query(...),
    status: Optional[str] = Query(default=None),
    priority_gte: Optional[int] = Query(default=None),
    due_before: Optional[datetime] = Query(default=None),
    assigned_only: bool = Query(default=False),
    case_severity_gte: Optional[str] = Query(default=None),
    sort: str = Query(default="priority"),
    db: Session = Depends(get_db),
):
    items = work_engine_service.list_work_items(
        db,
        business_id=business_id,
        status=status,
        priority_gte=priority_gte,
        due_before=due_before,
        assigned_only=assigned_only,
        case_severity_gte=case_severity_gte,
        sort=sort,
    )
    return {"items": items, "total": len(items)}


@router.post("/materialize", dependencies=[Depends(require_membership_dep(min_role="staff"))])
def post_materialize_work(
    case_id: str = Query(...),
    business_id: str = Query(...),
    db: Session = Depends(get_db),
):
    case = work_engine_service.case_engine_service._require_case(db, case_id)
    if case.business_id != business_id:
        raise HTTPException(status_code=404, detail="case not found")
    rows = work_engine_service.materialize_work_items_for_case(db, case_id)
    db.commit()
    return {"items": [{"id": row.id, "idempotency_key": row.idempotency_key, "status": row.status} for row in rows]}


@router.post("/{work_item_id}/complete", dependencies=[Depends(require_membership_dep(min_role="staff"))])
def post_complete_work_item(work_item_id: str, business_id: str = Query(...), db: Session = Depends(get_db)):
    existing = db.get(WorkItem, work_item_id)
    if not existing or existing.business_id != business_id:
        raise HTTPException(status_code=404, detail="work item not found")
    row = work_engine_service.complete_work_item(db, work_item_id)
    db.commit()
    return {"ok": True, "id": row.id, "status": row.status}


@router.post("/{work_item_id}/snooze", dependencies=[Depends(require_membership_dep(min_role="staff"))])
def post_snooze_work_item(work_item_id: str, req: SnoozeWorkIn, business_id: str = Query(...), db: Session = Depends(get_db)):
    existing = db.get(WorkItem, work_item_id)
    if not existing or existing.business_id != business_id:
        raise HTTPException(status_code=404, detail="work item not found")
    row = work_engine_service.snooze_work_item(db, work_item_id, snoozed_until=req.snoozed_until)
    db.commit()
    return {"ok": True, "id": row.id, "status": row.status, "snoozed_until": row.snoozed_until}

File: frontend/src/api/work.ts
import { apiGet, apiPost } from "./client";

export type WorkItemStatus = "open" | "snoozed" | "completed";

export type WorkItem = {
  id: string;
  case_id: string;
  business_id: string;
  type: string;
  priority: number;
  status: WorkItemStatus;
  due_at: string | null;
  snoozed_until: string | null;
  created_at: string;
  updated_at: string;
  resolved_at: string | null;
  idempotency_key: string;
  case_severity: string;
  case_domain: string;
  assigned_to: string | null;
};

export function listWorkItems(
  businessId: string,
  params?: {
    status?: WorkItemStatus;
    priority_gte?: number;
    due_before?: string;
    assigned_only?: boolean;
    case_severity_gte?: string;
    sort?: "priority" | "due_at" | "created_at";
  },
) {
  const query = new URLSearchParams({ business_id: businessId });
  if (params?.status) query.set("status", params.status);
  if (params?.priority_gte !== undefined) query.set("priority_gte", String(params.priority_gte));
  if (params?.due_before) query.set("due_before", params.due_before);
  if (params?.assigned_only !== undefined) query.set("assigned_only", String(params.assigned_only));
  if (params?.case_severity_gte) query.set("case_severity_gte", params.case_severity_gte);
  if (params?.sort) query.set("sort", params.sort);
  return apiGet<{ items: WorkItem[]; total: number }>(`/api/work?${query.toString()}`);
}

export function completeWorkItem(businessId: string, workItemId: string) {
  return apiPost(`/api/work/${encodeURIComponent(workItemId)}/complete?business_id=${encodeURIComponent(businessId)}`, {});
}

export function snoozeWorkItem(businessId: string, workItemId: string, snoozedUntil: string) {
  return apiPost(`/api/work/${encodeURIComponent(workItemId)}/snooze?business_id=${encodeURIComponent(businessId)}`, { snoozed_until: snoozedUntil });
}
